File: golang/316.go
func removeDuplicateLetters(s string) string {
    letter := make([]int, 26,26)
    for _, val := range s {
        letter[val-'a']++
    }
    visited := make([]bool, 26, 26)
    var ans []rune
    for _, val := range s{
        letter[val-'a']--
        if !visited[val-'a'] {
            for ;len(ans)>0 && ans[len(ans)-1] > val && letter[ans[len(ans)-1]-'a']>0; {
                visited[ans[len(ans)-1]-'a']= false
                ans[len(ans)-1] = 0
                ans = ans[:len(ans)-1]
            }
            visited[val-'a']=true
            ans = append(ans, val)
        }
    }
    return string(ans)
}

File: cpp/316.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        vector<int> letter(26,0);
        for(char i:s){
            letter[i-'a']++;
        }
        vector<bool>visited(26,0);
        string ans="";
        for(char i: s){
            letter[i-'a']--;
            if(!visited[i-'a']){
                while(ans.size() && ans.back()>i && letter[ans.back()-'a']){
                    visited[ans.back()-'a']=false;
                    ans.pop_back();
                }
                visited[i-'a']=true;
                ans+=i;
            }
        }
        return ans;
    }
};
